src/librustc/middle/mem_categorization.rs

```diff
@@ -1115,138 +634,71 @@ impl<'a, 'tcx> MemCategorizationContext<'a, 'tcx> {
-                self.cat_pattern_(cmt, &subpat, op)?;
+                self.cat_pattern_(place, &subpat, op)?;
             }
 
-            PatKind::Tuple(ref subpats, ddpos) => {
+            PatKind::Tuple(ref subpats, _) => {
                 // (p1, ..., pN)
-                let ty = self.pat_ty_unadjusted(&pat)?;
-                let expected_len = match ty.kind {
-                    ty::Tuple(ref tys) => tys.len(),
-                    _ => span_bug!(pat.span, "tuple pattern unexpected type {:?}", ty),
-                };
-                for (i, subpat) in subpats.iter().enumerate_and_adjust(expected_len, ddpos) {
-                    let subpat_ty = self.pat_ty_adjusted(&subpat)?; // see (*2)
-                    let interior = InteriorField(FieldIndex(i, sym::integer(i)));
-                    let subcmt = Rc::new(
-                        self.cat_imm_interior(pat, cmt.clone(), subpat_ty, interior));
-                    self.cat_pattern_(subcmt, &subpat, op)?;
+                for subpat in subpats.iter() {
+                    let subpat_ty = self.pat_ty_adjusted(&subpat)?;
+                    let subplace = self.cat_projection(pat, place.clone(), subpat_ty);
+                    self.cat_pattern_(subplace, &subpat, op)?;
                 }
             }
 
             PatKind::Box(ref subpat) | PatKind::Ref(ref subpat, _) => {
                 // box p1, &p1, &mut p1.  we can ignore the mutability of
                 // PatKind::Ref since that information is already contained
                 // in the type.
-                let subcmt = Rc::new(self.cat_deref(pat, cmt, NoteNone)?);
-                self.cat_pattern_(subcmt, &subpat, op)?;
+                let subplace = self.cat_deref(pat, place)?;
+                self.cat_pattern_(subplace, &subpat, op)?;
             }
 
             PatKind::Slice(ref before, ref slice, ref after) => {
-                let element_ty = match cmt.ty.builtin_index() {
+                let element_ty = match place.ty.builtin_index() {
                     Some(ty) => ty,
                     None => {
-                        debug!("explicit index of non-indexable type {:?}", cmt);
+                        debug!("explicit index of non-indexable type {:?}", place);
                         return Err(());
                     }
                 };
-                let context = InteriorOffsetKind::Pattern;
-                let elt_cmt = Rc::new(self.cat_index(pat, cmt, element_ty, context)?);
+                let elt_place = self.cat_projection(pat, place, element_ty);
                 for before_pat in before {
-                    self.cat_pattern_(elt_cmt.clone(), &before_pat, op)?;
+                    self.cat_pattern_(elt_place.clone(), &before_pat, op)?;
                 }
                 if let Some(ref slice_pat) = *slice {
-                    self.cat_pattern_(elt_cmt.clone(), &slice_pat, op)?;
+                    // FIXME the type here is wrong.
```

Comment:
Can you elaborate on this? Does it impact stabilizing slice_patterns in any way?

Reply:
The type should be an array/slice type fpr slice patterns, not the element type. I've fixed this now, so no.